src/ripple/server/impl/BaseWSPeer.h

```diff
@@ -308,11 +341,38 @@ on_write_fin(error_code const& ec)
     if(ec)
         return fail(ec, "write_fin");
     wq_.pop_front();
+
+    if(do_close_)
+    {
+        impl().ws_.async_close(cr_, strand_.wrap(std::bind(
+            &BaseWSPeer::on_close, impl().shared_from_this(),
+                beast::asio::placeholders::error)));
+        return;
+    }
+
+    write_pending_ = false;
+    if(! wq_.empty() || (timer_action_ == 2))
+        on_write({});
+}
+
+template<class Handler, class Impl>
+void
+BaseWSPeer<Handler, Impl>::
+on_write_ping(error_code const& ec)
+{
+    if(ec)
+        return fail(ec, "ping_sent");
+
     if(do_close_)
+    {
         impl().ws_.async_close(cr_, strand_.wrap(std::bind(
             &BaseWSPeer::on_close, impl().shared_from_this(),
                 beast::asio::placeholders::error)));
-    else if(! wq_.empty())
+        return;
+    }
+
+    write_pending_ = false;
+    if(! wq_.empty())
```

Comment:
I don't know how `timer_action_` could already be 2 again here, but would it be worth checking?